src/main/java/com/gregtechceu/gtceu/api/gui/widget/directional/handlers/CoverableConfigHandler.java

```diff
@@ -0,0 +1,205 @@
+            .setChangeListener(this::coverItemChanged)
+            .setBackgroundTexture(new GuiTextureGroup(GuiTextures.SLOT, GuiTextures.IO_CONFIG_COVER_SLOT_OVERLAY)));
+        group.addWidget(new PredicatedButtonWidget(0, 0, 18, 18, CONFIG_BTN_TEXTURE, this::toggleConfigTab)
+            .setPredicate(() -> side != null && coverBehavior != null && machine.getCoverAtSide(side) instanceof IUICover));
+
+        checkCoverBehaviour();
+
+        return group;
+    }
+
+    private void coverItemChanged() {
+        closeConfigTab();
+
+        if (!(panel.getGui().entityPlayer instanceof ServerPlayer serverPlayer) || side == null)
+            return;
+
+        var item = transfer.getStackInSlot(0);
+        if (machine.getCoverAtSide(side) != null) {
+            machine.removeCover(false, side, serverPlayer);
+        }
+
+        if (!item.isEmpty() && machine.getCoverAtSide(side) == null) {
+            if (item.getItem() instanceof ComponentItem componentItem) {
+                for (IItemComponent component : componentItem.getComponents()) {
+                    if (component instanceof CoverPlaceBehavior placeBehavior) {
+                        machine.placeCoverOnSide(side, item, placeBehavior.coverDefinition(), serverPlayer);
+                        break;
+                    }
+                }
+            }
+        }
+
+        checkCoverBehaviour();
+    }
+
+    @Override
+    public void onSideSelected(BlockPos pos, Direction side) {
+        this.side = side;
+        checkCoverBehaviour();
+        closeConfigTab();
+    }
+
+    private void updateWidgetVisibility() {
+        var sideSelected = this.side != null;
+        slotWidget.setVisible(sideSelected);
+        slotWidget.setActive(sideSelected);
+    }
+
+    public void checkCoverBehaviour() {
+        if (side == null)
+            return;
+
+        var coverBehaviour = machine.getCoverAtSide(side);
+        if (coverBehaviour != this.coverBehavior) {
+            this.coverBehavior = coverBehaviour;
+
+            var attachItem = coverBehaviour == null ? ItemStack.EMPTY : coverBehaviour.getAttachItem();
+            transfer.setStackInSlot(0, attachItem);
+            transfer.onContentsChanged(0);
+        }
+
+        updateWidgetVisibility();
+    }
+
+    private void toggleConfigTab(ClickData cd) {
+        if (this.coverConfigurator == null)
+            openConfigTab();
+        else
+            closeConfigTab();
+    }
+
+    private void openConfigTab() {
+        CoverConfigurator configurator = new CoverConfigurator(this.machine, this.side, this.coverBehavior) {
+            @Override
+            public Component getTitle() {
+                // Uses the widget's own title
+                return Component.empty();
+            }
+
+            @Override
+            public IGuiTexture getIcon() {
+                // TODO make a proper texture for this
+                return new TextTexture("x", Color.BLACK.getRGB());
```

Comment:
DO NOT USE ANYTHING FROM AWT it isn't shipped in some binaries post-java 8

Reply:
Oh thanks lol, didn't realize this was AWT